### ISSUE TITLE
Is there a demo for parallel BO?

### ISSUE THREAD

**Issue by deng-cy** [pytorch/botorch#1356]
Since the evaluation of my objective function takes a lot of time, I am trying to use parallel BO to run several trials simultaneously. Yet I did not find any demo on how to use parallel BO. I know Ax could do this without much effort, but there are some other issues such that I cannot use Ax. I am wondering whether there is any instruction to let me know how to use parallel BO purely by BoTorch.

**Comment by eytan** [comment]
https://botorch.org/tutorials/closed_loop_botorch_only

Gives an example of how to do BO in parallel. Here the parallelism is
controlled by BATCH_SIZE. If you are having difficulties with ax we’d be
happy to help if you file an issue on the ax GitHub repo.

**Comment by deng-cy** [comment]
Thanks for the reply! One thing slightly different from what I want is that this parallelization is not sequential. I hope to have something similar to the service API of ax. For instance, there are 5 parallel workers, and as long as one of them is done, I can get a new candidate immediately without waiting for the other 4. Yet the BATCH_SIZE (or q) does not seem to be able to handle this. 

I'll post an issue on ax GitHub soon.

**Comment by saitcakmak** [comment]
Hi @deng-cy. What you're looking for is the `X_pending` argument of the MC acquisition functions.
```
X_pending: A `batch_shape, m x d`-dim Tensor of `m` design points
                that have points that have been submitted for function evaluation
                but have not yet been evaluated.
```
In don't think we have a tutorial demonstrating the asynchronous batch evaluation setup in BoTorch. This generally requires use of a scheduler to coordinate evaluation of the function, for which we recommend using Ax. 

Regardless, if you wanted to do this purely in BoTorch, here's what you'd need to do. Lets call `K` the # parallel jobs you want to have running.
- Assuming you start with `K` of jobs running in parallel, either generated manually, via Sobol, or using `q=K` in `optimize_acqf` with the standard setup. Do the following in a loop.
- Wait for one or more of the jobs to complete. Suppose `k` of them completed. Add both `X` and `Y` s of these `k` jobs to your model training data and re-train the model.
- Define `X_pending` to be a `(K - k) x d`-dim tensor representing the `X`s of `K - k` jobs still being evaluated.
- Construct the acquisition function as usual, including `X_pending=X_pending` in its arguments. So, something like `acqf = qNoisyExpectedImprovement(model=model, X_baseline=train_X, X_pending=X_pending)`.
- Optimize the acquisition function to generate `k` candidates to bring the total of running jobs to `K`. So, `optimize_acqf(acq_function=acqf, q=k, ...)`.
- Submit the `k` candidates you just generated for evaluation. You should have a total of `K` jobs running in parallel again.

**Comment by deng-cy** [comment]
Great! Thank you so much.
